sklearn/model_selection/_classification_threshold.py

```diff
@@ -0,0 +1,1189 @@
+        already fitted.
+
+    val_idx : ndarray of shape (n_val_samples,)
+        The indices of the validation set used to score `classifier`. If `train_idx`,
+        the entire set will be used.
+
+    curve_scorer : scorer instance
+        The scorer taking `classifier` and the validation set as input and outputting
+        decision thresholds and scores as a curve. Note that this is different from
+        the usual scorer that output a single score value:
+
+        * when `score_method` is one of the four constraint metrics, the curve scorer
+          will output a curve of two scores parametrized by the decision threshold, e.g.
+          TPR/TNR or precision/recall curves for each threshold;
+        * otherwise, the curve scorer will output a single score value for each
+          threshold.
+
+    score_params : dict
+        Parameters to pass to the `score` method of the underlying scorer.
+
+    Returns
+    -------
+    potential_thresholds : ndarray of shape (thresholds,)
+        The decision thresholds used to compute the scores. They are returned in
+        ascending order.
+
+    scores : ndarray of shape (thresholds,) or tuple of such arrays
+        The scores computed for each decision threshold. When TPR/TNR or precision/
+        recall are computed, `scores` is a tuple of two arrays.
+    """
+
+    if train_idx is not None:
+        X_train, X_val = _safe_indexing(X, train_idx), _safe_indexing(X, val_idx)
+        y_train, y_val = _safe_indexing(y, train_idx), _safe_indexing(y, val_idx)
+        fit_params_train = _check_method_params(X, fit_params, indices=train_idx)
+        score_params_val = _check_method_params(X, score_params, indices=val_idx)
+        classifier.fit(X_train, y_train, **fit_params_train)
+    else:  # prefit estimator, only a validation set is provided
+        X_val, y_val, score_params_val = X, y, score_params
+
+    if curve_scorer is roc_curve or (
+        isinstance(curve_scorer, _BaseScorer) and curve_scorer._score_func is roc_curve
+    ):
+        fpr, tpr, potential_thresholds = curve_scorer(
+            classifier, X_val, y_val, **score_params_val
+        )
+        # For fpr=0/tpr=0, the threshold is set to `np.inf`. We need to remove it.
+        fpr, tpr, potential_thresholds = fpr[1:], tpr[1:], potential_thresholds[1:]
+        # thresholds are in decreasing order
+        return potential_thresholds[::-1], ((1 - fpr)[::-1], tpr[::-1])
+    elif curve_scorer is precision_recall_curve or (
+        isinstance(curve_scorer, _BaseScorer)
+        and curve_scorer._score_func is precision_recall_curve
+    ):
+        precision, recall, potential_thresholds = curve_scorer(
+            classifier, X_val, y_val, **score_params_val
+        )
+        # thresholds are in increasing order
+        # the last element of the precision and recall is not associated with any
+        # threshold and should be discarded
+        return potential_thresholds, (precision[:-1], recall[:-1])
```

Comment:
What I mean is to have a new `CureScorerMixin` and have `roc_auc` and `precision_recall_curve` to be one such scorer as well.

Reply:
As I mentioned earlier I also think that there's room for improvement in the Scorers design to simplify this, but it is the kind of internal details that I think can be sorted out in a follow-up PR because it doesn't impact the public API.

Reply:
Indeed, we had this exact discussion with @jeremiedbb as he mentioned we thought about not addressing it in this PR.

Reply:
It is the kind of technical debt that we don't tend to pay immediately after, that's why I'm worried. Could you please open an issue for this?

Reply:
+1 for creating an issue, after merging it because it's hard to point to code that does not exist yet 😃

Reply:
We can point to this PR

Reply:
I open the following issue to be edited in the future: https://github.com/scikit-learn/scikit-learn/issues/28941